calculator/calculator.py

```diff
@@ -0,0 +1,218 @@
+
+    Attributes:
+        operation (function): the function that does the operation.
+        precedence (int): the priority of the operation.
+        regex (str): regex to find a pattern which suits to the operator.
+
+    Note:
+        Precedence determines which operation goes after another.
+    """
+
+    def __init__(self, operation, precedence, regex):
+        self.operation = operation
+        self.precedence = precedence
+        self.regex = re.compile(regex)
+
+
+def average(num1, num2):
+    """Return the average of two numbers (int/float).
+
+    Note:
+        Same as using mean() from statistics module.
+    """
+    return float(num1 + num2) / 2.0
+
+
+def remove_spaces(expression: str) -> str:
+    """Remove the spaces between characters in a string."""
+    return ''.join(expression.split())
+
+
+def get_sub_expression_by_symbol(symbol: str, expression: str) -> str:
+    """Find the sub expression the current operator.
+
+    Args:
+        symbol: the current operator symbol.
+        expression: the entire expression to search.
+
+    Returns:
+         The desired sub expression of the operator. If nothing is found
+         returns an empty string.
+    """
+    match = operators[symbol].regex.search(expression)
+
+    if not match:
+        return ""
+
+    return match.group()
+
+
+def check_operands(func):
+    """Check the existence of the operands.
+
+    Args:
+        func (function): the function to operate on.
+
+    Returns:
+        function: the function wrapper.
+    """
+
+    def operands_wrapper(operator_symbol, operands):
+        left_operand, right_operand = operands
+
+        if left_operand and right_operand:
+            return func(operator_symbol, operands)
+
+        elif not left_operand and right_operand:
+            # if any([operator_symbol == "+", operator_symbol == "-"]):
+            #     return float(f"{operator_symbol}{right_operand}")
+            return func(operator_symbol, [right_operand])
+
+        elif not right_operand and left_operand:
+            return func(operator_symbol, [left_operand])
+
+    return operands_wrapper
+
+
+@check_operands
+def calculate_sub_expression(symbol: str, operands: list) -> float:
+    """Evaluate an expression.
+
+    Args:
+        operands (list): the operands, can be 2 or 1 according to operation.
+        symbol (str): a symbol represents the mathematical operation.
+
+    Returns:
+        A float, the result of the mathematical operation.
+
+    Note:
+        Brackets need a string operand (_evaluate() takes a string) and the
+        others a float (factorial and negate).
+    """
+    if len(operands) == 2:
+        return operators[symbol].operation(float(operands[0]),
+                                           float(operands[1]))
+
+    operand = operands[0]
+    operation_to_do = operators[symbol].operation
+
+    if operation_to_do is not _evaluate:
+        operand = float(operands[0])
+
+    return operation_to_do(operand)
```

Comment:
What if want to support more not only binary operators? 
Make sure your code can support n-nary operators :)